### PR TITLE
python/isbn-verifier: add mentor notes

### PR DESCRIPTION
**Pull request by IsaacG**

### CODE REVIEW

**Comment by IsaacG** [comment]
@BethanyG for review, please.

File: tracks/python/exercises/isbn-verifier/mentoring.md
Comment (kotp) [inline tracks/python/exercises/isbn-verifier/mentoring.md:75]:
```suggestion
Recognize that changing the data from `str` to `list[str]` allows for more flexibility of what can be stored (`"10"`) and a simpler summation loop.
```

File: tracks/python/exercises/isbn-verifier/mentoring.md
Comment (kotp) [inline tracks/python/exercises/isbn-verifier/mentoring.md:79]: It is very convenient for any non-digit values in the input, but why?  How is it convenient?  What does it do for it?  It feels like there might be a word missing or perhaps a couple of words in this sentence.

File: tracks/python/exercises/isbn-verifier/mentoring.md
Comment (kotp) [inline tracks/python/exercises/isbn-verifier/mentoring.md:86]: I tend to shy away from a double negative in the conditional statement, is there another way to express this with out the not and not equal?

File: tracks/python/exercises/isbn-verifier/mentoring.md
Comment (kotp) [inline tracks/python/exercises/isbn-verifier/mentoring.md:107]:
```suggestion
Checking for non-digits separately from the summation allows for explicitly and directly handling that test.
```

**Review by kotp — CHANGES_REQUESTED** [review]

File: tracks/python/exercises/isbn-verifier/mentoring.md
Comment (IsaacG) [reply tracks/python/exercises/isbn-verifier/mentoring.md:11]: This was intended as pseudo-code. Or, as a compact and succient manner to encapsulate the requirements here. I could write it out in English in a long form.

File: tracks/python/exercises/isbn-verifier/mentoring.md
Comment (IsaacG) [reply tracks/python/exercises/isbn-verifier/mentoring.md:36]: The `sum()` generator is in the "Reasonable Solution" section. Should it be repeated here? Referenced?

File: tracks/python/exercises/isbn-verifier/mentoring.md
Comment (IsaacG) [reply tracks/python/exercises/isbn-verifier/mentoring.md:45]: I could run benchmarks and dump values.
I could link to PyCon talks.
But this is guidance for mentors, like a research paper :)

Are you referring specifically to line 45? Or the entire argument in favor of using generators?

File: tracks/python/exercises/isbn-verifier/mentoring.md
Comment (IsaacG) [reply tracks/python/exercises/isbn-verifier/mentoring.md:49]: Adjusted.

File: tracks/python/exercises/isbn-verifier/mentoring.md
Comment (IsaacG) [reply tracks/python/exercises/isbn-verifier/mentoring.md:79]: What would you like to see here? How about:
```suggestion
The [built-in `any()` function][builtin-functions] is a concise approach for checking for any non-digit values in the input.
```

File: tracks/python/exercises/isbn-verifier/mentoring.md
Comment (IsaacG) [reply tracks/python/exercises/isbn-verifier/mentoring.md:11]:
```suggestion
* The ISBN is valid if the weighted sum of the digits is a multiple of 11 where the weight is the digit's value, ranging from 10 to 1.
```

File: tracks/python/exercises/isbn-verifier/mentoring.md
Comment (IsaacG) [reply tracks/python/exercises/isbn-verifier/mentoring.md:49]:
```suggestion
The position values can be directly generated using `range()`.
```

File: tracks/python/exercises/isbn-verifier/mentoring.md
Comment (IsaacG) [reply tracks/python/exercises/isbn-verifier/mentoring.md:86]: This isn't a double negative as the two negatives are applying to different conditionals. Do you prefer this?

```suggestion
if isbn[-1].isdigit() or isbn[-1] = "X":
    pass
else:
```

File: tracks/python/exercises/isbn-verifier/mentoring.md
Comment (kotp) [reply tracks/python/exercises/isbn-verifier/mentoring.md:36]: You could cross reference it:  "See the Reasonable Solution section." for example.

File: tracks/python/exercises/isbn-verifier/mentoring.md
Comment (kotp) [reply tracks/python/exercises/isbn-verifier/mentoring.md:45]: Specifically "They are also very efficient."  This could be "This is known to be more efficient than other potential approaches." Maybe.  But if there are references it is good to provide further reading to support this statement.  Otherwise we could pose it as a question for exploration.

File: tracks/python/exercises/isbn-verifier/mentoring.md
Comment (kotp) [reply tracks/python/exercises/isbn-verifier/mentoring.md:79]: That makes more sense to me.

File: tracks/python/exercises/isbn-verifier/mentoring.md
Comment (kotp) [reply tracks/python/exercises/isbn-verifier/mentoring.md:86]: Yes, I prefer that, it is a more positive thing.  It also has the different body, from `return False` to just a `pass`.  Surprising to me, in a way, as I would have thought the other potential would have been `True`.  Because with the else, that should be the true branch, and I prefer true and then false conditional branches to return true and then false.  Otherwise I usually consider the conditional statement to be "backward".

File: tracks/python/exercises/isbn-verifier/mentoring.md
Comment (kotp) [inline tracks/python/exercises/isbn-verifier/mentoring.md:36]:
```suggestion
The sum can be computed using a `for` loop or via the `sum()` [built-in function][builtin-functions].
```

**Review by kotp — CHANGES_REQUESTED** [review]

File: tracks/python/exercises/isbn-verifier/mentoring.md
Comment (kotp) [reply tracks/python/exercises/isbn-verifier/mentoring.md]: I think this is better than the former version.  Otherwise use variable names, and then explain what those variable names reference.  Programmer are used to interpreting variable names (especially if they are well named) and inferring what is meant.

But I like the written out explanation.  I can see the value of having the code presented to show the applied description.  

So, both?

File: tracks/python/exercises/isbn-verifier/mentoring.md
Comment (IsaacG) [reply tracks/python/exercises/isbn-verifier/mentoring.md:45]: Couching it in more vague language shifts it from a "claim from authority" to an "appeal to the people". You're saying the same thing ... but saying "it's known" vs "I know". It doesn't change the substance in any way.

Generators store less state than iterators and require fewer lookups, which is why they're faster than iterators. Generators don't build a list (which requires list appends and storing all the elements) which is why they are faster than building a list. Should I include these details?

There is PEP 255 where generators are introduced; the PEP lays out the reason for generators which includes the inefficiencies of lists and iterators for this use case. Generators are introduced in part *because* they are more efficient. I suppose the PEP itself is a pretty good reference for their efficiency, if someone wanted to dig through it. (See Motivation, paragraphs 3 and 4.)

Or I could just state they are more efficient ;)\

> Otherwise we could pose it as a question for exploration

Are we posing the question to mentors? Suggesting mentors suggest students benchmark it?

References:

https://www.dabeaz.com/generators/ and his PyCon talks are an amazing watch/listen on generators. Googling for "python generators vs iterators" produces numerous results. https://www.python.org/dev/peps/pep-0255/ motivation.

See https://medium.com/analytics-vidhya/python-iterators-vs-iterables-vs-generators-e15b9403292e

> Generator is efficient for writing fast and compact code. This is an advantage over Python iterators. They are also simpler to code than do custom iterator.

File: tracks/python/exercises/isbn-verifier/mentoring.md
Comment (IsaacG) [reply tracks/python/exercises/isbn-verifier/mentoring.md:86]: This is one of the many tests required to check if a string is a valid ISBN. All the tests need to pass before we can `return True`. I would argue that

```
if not condition:
    return False
```
is significantly more compact, readable and Pythonic than:
```
if condition:
    pass
else:
    return False
```

What if I wrapped it all in a single `not`?

```suggestion
if not (isbn[-1].isdigit() or isbn[-1] == "X"):
```

File: tracks/python/exercises/isbn-verifier/mentoring.md
Comment (IsaacG) [inline tracks/python/exercises/isbn-verifier/mentoring.md]:
```suggestion
The sum can be computed using the `sum()` [built-in function][builtins-functions] (see "Reasonable Solution") or a `for` loop.
```

Comment (IsaacG) [reply tracks/python/exercises/isbn-verifier/mentoring.md]: Done.

File: tracks/python/exercises/isbn-verifier/mentoring.md
Comment (IsaacG) [reply tracks/python/exercises/isbn-verifier/mentoring.md]: If you want working code which demonstrates what is needed ... isn't that what the next section contains?

Comment (kotp) [reply tracks/python/exercises/isbn-verifier/mentoring.md]: Then, yes, prefer the written here and the example in the next section.

Comment (IsaacG) [reply tracks/python/exercises/isbn-verifier/mentoring.md]: Done in this current iteration.

File: tracks/python/exercises/isbn-verifier/mentoring.md
Comment (kotp) [reply tracks/python/exercises/isbn-verifier/mentoring.md:45]: So I think you are saying that you could reference PEP 255 where efficiencies are mentioned.  (More efficient in what?  CPU ticks, Memory, Power, Time complexity?) Or perhaps the references you show here are good items to show in the mentor notes?

> Are we posing the question to mentors?

This document can supply questions to pose to students, sometimes as copy/paste-able material.  And usually I avoid benchmark suggestions, unless the benchmarks are extremely simple, or if the student brings it up, and even them usually we find out if that is really the bottleneck in their solution, often by profiling instead, or at least before we benchmark.

File: tracks/python/exercises/isbn-verifier/mentoring.md
Comment (kotp) [reply tracks/python/exercises/isbn-verifier/mentoring.md:86]: With the following line being `return False` I do question if that is correct, as the last "digit" is allowed to be "X".

It is part of the larger `any()` statement, so I will defer to you for the specifics here.  It is only that I wanted to consider options here.

Comment (IsaacG) [reply tracks/python/exercises/isbn-verifier/mentoring.md]: This line is separate from the `any()` statement. The last character of the ISBN is allowed to be `"X"`, per the problem spec.

File: tracks/python/exercises/isbn-verifier/mentoring.md
Comment (IsaacG) [reply tracks/python/exercises/isbn-verifier/mentoring.md]: Reference to PEP-255 added.

File: tracks/python/exercises/isbn-verifier/mentoring.md
Comment (kotp) [inline tracks/python/exercises/isbn-verifier/mentoring.md:45]:
```suggestion
Generators were added to the language (see [PEP-255]) because they provide a compact, clean and efficient method for iterating through data (compared to other approaches such as building lists or iterators).
```

**Review by kotp — DISMISSED** [review]

**Comment by IsaacG** [comment]
Would you mind doing a (squash and) merge, too? Or are we waiting for Bethany to take a look?

**Comment by kotp** [comment]
The contributor should be responsible for what their history looks like.  So if you want to change the history and force push to this branch, please do so, have the history look like you want it to be told.

**Comment by IsaacG** [comment]
I don't have permissions to merge in this repo. I can squash and force update, though. Will do that.

**Review by kotp — APPROVED** [review]
LGTM

**Comment by IsaacG** [comment]
This is good to be merged from my side. But I don't have write access.

**Pull request merged**
